src/EmbeddingSpMDMAutovec.cc

```diff
@@ -435,6 +435,483 @@ bool EmbeddingSpMDMNBit_autovec(
   }
   return current == index_size;
 }
+template <
+    typename InType,
+    typename IndexType,
+    typename OffsetType,
+    typename OutType>
+bool EmbeddingSpMDM_autovec(
+    const int64_t block_size,
+    const int64_t output_size,
+    const int64_t index_size,
+    const int64_t data_size,
+    const InType* input,
+    const IndexType* indices,
+    const OffsetType* offsets_or_lengths,
+    const float* weights, // optional, can be null for non-weighted sum
+    bool normalize_by_lengths,
+    OutType* out,
+    bool is_weight_positional /*=false*/,
+    bool use_offsets /*=true*/,
+    int64_t output_stride /*=-1*/,
+    int64_t input_stride /*=-1*/,
+    bool scale_bias_last /*=true*/,
+    bool no_bag /*=false*/,
+    bool is_bf16_out /*=false*/,
+    bool is_bf16_in /*=false*/) {
+  const bool isWeight8bit = std::is_same<InType, uint8_t>::value;
+  const bool isOutput8bit = std::is_same<OutType, uint8_t>::value;
+  if (output_stride == -1) {
+    output_stride = block_size;
+  }
+  if constexpr (isOutput8bit) {
+    assert(input_stride == output_stride);
+  }
+  vector<float> buf(block_size);
+
+  if (isWeight8bit) {
+    // block_size is the number of elements and fused_block_size is the size of
+    // an entire row, including scale and bias.
+    if (input_stride == -1) {
+      // scale_bias_last == false is for table batched embedding that stores
+      // scale and bias in float16
+      const auto scale_bias_offset =
+          2 * (scale_bias_last ? sizeof(float) : sizeof(float16));
+      input_stride = block_size + scale_bias_offset;
+    }
+    int64_t current = 0;
+
+    if (no_bag) {
+      for (int m = 0; m < output_size; ++m) {
+        int64_t idx = indices[m];
+
+        if (idx < 0 || idx >= data_size) {
+          return false;
+        }
+        if constexpr (isOutput8bit) {
+          const InType* input_row_ptr = input + input_stride * idx;
+          memcpy(out, input_row_ptr, sizeof(InType) * input_stride);
+        } else {
+          memset(buf.data(), 0, sizeof(float) * block_size);
```

Comment:
Is initializing to zero and using fma helps vectorization vs. leave with garbage values and overwrite with multiplication results?

Reply:
Yeah initializing to zero is not really necessary for sequential embeddings because there is no aggregation from multiple embedding rows. It is required in pooled embeddings for correctness, though.